lib/irq.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2019, Xilinx Inc. and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <errno.h>
+#include <metal/irq.h>
+#include <metal/list.h>
+#include <metal/utilities.h>
+
+/** List of registered IRQ controller */
+static METAL_DECLARE_LIST(irq_cntrs);
+
+static int metal_irq_allocate(int irq_base, int irq_num)
+{
+	struct metal_list *node;
+	struct metal_irq_controller *cntr;
+	int irq_tocheck = irq_base, irq_end_tocheck;
+
+	if (irq_num == 0) {
+		return METAL_IRQ_ANY;
+	}
+	if (irq_tocheck == METAL_IRQ_ANY) {
+		irq_tocheck = 0;
+	}
+	irq_end_tocheck = irq_tocheck + irq_num;
+
+	metal_list_for_each(&irq_cntrs, node) {
+		int cntr_irq_base, cntr_irq_end;
+
+		cntr = metal_container_of(node,
+					  struct metal_irq_controller, node);
+		cntr_irq_base = cntr->irq_base;
+		cntr_irq_end = cntr_irq_base + cntr->irq_num;
+		if (irq_tocheck < cntr_irq_end &&
+		    irq_end_tocheck > cntr_irq_base) {
```

Comment:
irq_end_tocheck >= cntr_irq_base?

Reply:
it will be ">" as the "irq_end_tocheck" is "irq_tocheck + irq_num" but not "irq_tocheck + irq_num - 1"

Reply:
Emm, you are right.